crypto/hpke/hpke.c

```diff
@@ -0,0 +1,1350 @@
+                                  NULL, 0, OSSL_HPKE_SEC51LABEL,
+                                  suitebuf, sizeof(suitebuf),
+                                  OSSL_HPKE_PSK_HASH_LABEL,
+                                  ctx->psk, ctx->psklen) != 1) {
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+    secretlen = kdf_info->Nh;
+    if (secretlen > OSSL_HPKE_MAXSIZE) {
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+    if (ossl_hpke_labeled_extract(kctx, secret, secretlen,
+                                  ctx->shared_secret, ctx->shared_secretlen,
+                                  OSSL_HPKE_SEC51LABEL,
+                                  suitebuf, sizeof(suitebuf),
+                                  OSSL_HPKE_SECRET_LABEL,
+                                  ctx->psk, ctx->psklen) != 1) {
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+    if (ctx->suite.aead_id != OSSL_HPKE_AEAD_ID_EXPORTONLY) {
+        /* we only need nonce/key for non export AEADs */
+        ctx->noncelen = aead_info->Nn;
+        ctx->nonce = OPENSSL_malloc(ctx->noncelen);
+        if (ctx->nonce == NULL) {
+            goto err;
+        }
+        if (ossl_hpke_labeled_expand(kctx, ctx->nonce, ctx->noncelen,
+                                     secret, secretlen, OSSL_HPKE_SEC51LABEL,
+                                     suitebuf, sizeof(suitebuf),
+                                     OSSL_HPKE_NONCE_LABEL,
+                                     ks_context, ks_contextlen) != 1) {
+            ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+            goto err;
+        }
+        ctx->keylen = aead_info->Nk;
+        ctx->key = OPENSSL_malloc(ctx->keylen);
+        if (ctx->key == NULL) {
+            goto err;
+        }
+        if (ossl_hpke_labeled_expand(kctx, ctx->key, ctx->keylen,
+                                     secret, secretlen, OSSL_HPKE_SEC51LABEL,
+                                     suitebuf, sizeof(suitebuf),
+                                     OSSL_HPKE_KEY_LABEL,
+                                     ks_context, ks_contextlen) != 1) {
+            ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+            goto err;
+        }
+    }
+    ctx->exporterseclen = kdf_info->Nh;
+    ctx->exportersec = OPENSSL_malloc(ctx->exporterseclen);
+    if (ctx->exportersec == NULL)
+        goto err;
+    if (ossl_hpke_labeled_expand(kctx, ctx->exportersec, ctx->exporterseclen,
+                                 secret, secretlen, OSSL_HPKE_SEC51LABEL,
+                                 suitebuf, sizeof(suitebuf),
+                                 OSSL_HPKE_EXP_LABEL,
+                                 ks_context, ks_contextlen) != 1) {
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+    erv = 1;
+
+err:
+    OPENSSL_cleanse(ks_context, OSSL_HPKE_MAXSIZE);
```

Comment:
On failure should this cleanup things created here such as the nonce? (I know it frees them on cleanup).

Reply:
I think perhaps it's better to assume the caller will call ``OSSL_HPKE_CTX_free()`` tbh (which will zap the possibly sensitive values, as you say). The alternative (cleaning up half-done stuff) might be more likely to lead to bugs over time. (But, I could do it if it's better.)